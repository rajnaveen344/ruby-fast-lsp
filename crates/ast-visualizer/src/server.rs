use actix_web::http::header::ContentType;
use actix_web::{web, HttpResponse, Responder, Result};
use ruby_prism::parse;
use serde::{Deserialize, Serialize};
use std::fs;
use std::path::Path;
use std::sync::Mutex;

use crate::ast::convert_node_to_ast;
use crate::utils::offset_to_position;

// Request structure for parsing Ruby code
#[derive(Deserialize, Serialize)]
pub struct ParseRequest {
    pub code: String,
}

// State to track request count
pub struct AppState {
    pub request_count: Mutex<usize>,
}

// Parse Ruby code and return AST as JSON
pub async fn parse_ruby(
    data: web::Json<ParseRequest>,
    state: web::Data<AppState>,
) -> impl Responder {
    // Increment request count - handle poisoned mutex gracefully
    let count_result = state.request_count.lock();
    let request_num = match count_result {
        Ok(mut count) => {
            *count += 1;
            *count
        }
        Err(poisoned) => {
            // If the mutex is poisoned, recover the lock and continue
            let mut count = poisoned.into_inner();
            *count += 1;
            *count
        }
    };
    println!("Request #{}: Parsing Ruby code", request_num);

    // Parse Ruby code
    let code = data.code.clone();

    // Check input size - limit to 500KB to prevent memory issues
    const MAX_INPUT_SIZE: usize = 500 * 1024; // 500KB
    if code.len() > MAX_INPUT_SIZE {
        return HttpResponse::BadRequest().json(serde_json::json!({
            "error": "Input too large",
            "message": format!("Input size ({} bytes) exceeds maximum allowed size ({} bytes)", code.len(), MAX_INPUT_SIZE)
        }));
    }

    // Validate input content
    // Check for non-UTF8 characters or other problematic content
    if code.contains("\0") {
        return HttpResponse::BadRequest().json(serde_json::json!({
            "error": "Invalid input",
            "message": "Input contains null bytes which are not allowed"
        }));
    }

    // Check for extremely long lines that might cause issues
    const MAX_LINE_LENGTH: usize = 10000; // 10K characters per line max
    if code.lines().any(|line| line.len() > MAX_LINE_LENGTH) {
        return HttpResponse::BadRequest().json(serde_json::json!({
            "error": "Invalid input",
            "message": format!("Input contains lines exceeding the maximum allowed length of {} characters", MAX_LINE_LENGTH)
        }));
    }

    // Safely parse the code with error handling
    let parse_result = match std::panic::catch_unwind(|| {
        // Use a more defensive approach to parsing
        parse(code.as_bytes())
    }) {
        Ok(result) => result,
        Err(_) => {
            return HttpResponse::InternalServerError().json(serde_json::json!({
                "error": "Parsing error",
                "message": "Failed to parse the input. The code may be too complex or contain unsupported constructs."
            }));
        }
    };

    // Check for syntax errors
    let errors_count = parse_result.errors().count();

    // If there are too many errors, reject the input
    const MAX_ERRORS: usize = 50;
    if errors_count > MAX_ERRORS {
        return HttpResponse::BadRequest().json(serde_json::json!({
            "error": "Too many syntax errors",
            "message": format!("The input contains {} syntax errors, which exceeds the maximum allowed ({})", errors_count, MAX_ERRORS)
        }));
    }

    // If there are some errors, return them in a structured format
    if errors_count > 0 {
        // Create a structured error response that can be visualized
        let error_nodes: Vec<serde_json::Value> = parse_result
            .errors()
            .map(|e| {
                let location = e.location();
                let message = format!("{:?}", e);

                // Get location information
                let start_offset = location.start_offset();
                let _end_offset = location.end_offset(); // We could use this for error highlighting

                // Calculate line and column from the code string
                let (start_line, start_column) = offset_to_position(&code, start_offset);

                serde_json::json!({
                    "type": "SyntaxErrorNode",
                    "message": message,
                    "line": start_line,
                    "column": start_column
                })
            })
            .collect();

        // Return a special AST structure for errors
        return HttpResponse::Ok().json(serde_json::json!({
            "type": "ErrorTreeNode",
            "name": "Syntax Errors",
            "children": error_nodes
        }));
    }

    // Convert AST to JSON-serializable structure with error handling
    let ast_result = std::panic::catch_unwind(|| convert_node_to_ast(&parse_result.node()));

    match ast_result {
        Ok(ast) => HttpResponse::Ok().json(ast),
        Err(_) => HttpResponse::InternalServerError().json(serde_json::json!({
            "error": "AST processing error",
            "message": "Failed to process the AST. The input may be too complex or contain unsupported constructs."
        }))
    }
}

// Health check endpoint
pub async fn health_check(state: web::Data<AppState>) -> impl Responder {
    // Handle poisoned mutex gracefully
    let count_result = state.request_count.lock();
    let count = match count_result {
        Ok(count) => *count,
        Err(poisoned) => {
            // If the mutex is poisoned, recover the lock and continue
            *poisoned.into_inner()
        }
    };

    HttpResponse::Ok().json(serde_json::json!({
        "status": "ok",
        "requests_processed": count
    }))
}

// Serve the HTML page
pub async fn index() -> Result<HttpResponse> {
    // Path to the HTML file
    let html_path = Path::new("static/index.html");

    // Check if the file exists in the current directory
    if html_path.exists() {
        let html_content = fs::read_to_string(html_path).map_err(|e| {
            eprintln!("Error reading HTML file: {}", e);
            actix_web::error::ErrorInternalServerError("Error reading HTML file")
        })?;

        Ok(HttpResponse::Ok()
            .content_type(ContentType::html())
            .body(html_content))
    } else {
        // Try to find the file relative to the crate directory
        let crate_html_path = Path::new("crates/ast-visualizer/static/index.html");

        if crate_html_path.exists() {
            let html_content = fs::read_to_string(crate_html_path).map_err(|e| {
                eprintln!("Error reading HTML file: {}", e);
                actix_web::error::ErrorInternalServerError("Error reading HTML file")
            })?;

            Ok(HttpResponse::Ok()
                .content_type(ContentType::html())
                .body(html_content))
        } else {
            eprintln!(
                "HTML file not found at {:?} or {:?}",
                html_path, crate_html_path
            );
            Ok(HttpResponse::NotFound().body("HTML file not found"))
        }
    }
}
