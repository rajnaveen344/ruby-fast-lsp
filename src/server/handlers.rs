use anyhow::Result;
use log::{info, warn};
use lsp_types::{
    CompletionResponse, Hover, HoverContents,
    MarkupContent, MarkupKind, Position, Range,
};
use tower_lsp::jsonrpc::Result as LspResult;

use crate::analysis::RubyAnalyzer;
use crate::parser::{document::RubyDocument, RubyParser};

#[derive(Clone)]
pub struct RubyLspHandlers {
    parser: Option<RubyParser>,
    analyzer: RubyAnalyzer,
}

impl RubyLspHandlers {
    pub fn new() -> Result<Self> {
        let parser = match RubyParser::new() {
            Ok(parser) => Some(parser),
            Err(e) => {
                warn!("Failed to initialize Ruby parser: {}", e);
                None
            }
        };
        
        Ok(Self {
            parser,
            analyzer: RubyAnalyzer::new(),
        })
    }

    pub fn handle_hover(&self, document: &RubyDocument, position: Position) -> Option<Hover> {
        info!("Handling hover request at position {:?}", position);
        
        let tree = self.parse_document(document);
        
        // Get hover information
        let hover_info = self
            .analyzer
            .get_hover_info(tree.as_ref(), document.get_content(), position)
            .unwrap_or_else(|| "No information available".to_string());
        
        Some(Hover {
            contents: HoverContents::Markup(MarkupContent {
                kind: MarkupKind::Markdown,
                value: hover_info,
            }),
            range: None,
        })
    }

    pub fn handle_completion(
        &self,
        document: &RubyDocument,
        position: Position,
    ) -> CompletionResponse {
        info!("Handling completion request at position {:?}", position);
        
        let tree = self.parse_document(document);
        
        // Get completions from the analyzer
        let items = self.analyzer.get_completions(tree.as_ref(), document.get_content(), position);
        
        CompletionResponse::Array(items)
    }

    pub fn handle_definition(
        &self,
        document: &RubyDocument,
        position: Position,
    ) -> Option<Range> {
        info!("Handling definition request at position {:?}", position);
        
        let tree = self.parse_document(document);
        
        // Find definition
        self.analyzer.find_definition(tree.as_ref(), document.get_content(), position)
    }
    
    // Helper method to parse a document and return the tree
    fn parse_document(&self, document: &RubyDocument) -> Option<tree_sitter::Tree> {
        // If parser is not available, return None
        let parser = match &self.parser {
            Some(parser) => parser,
            None => {
                warn!("Parser not available for document parsing");
                return None;
            }
        };
        
        // Try to parse the document
        match parser.parse(document.get_content()) {
            Some(tree) => {
                info!("Document parsed successfully");
                Some(tree)
            },
            None => {
                warn!("Failed to parse document");
                None
            }
        }
    }
}
