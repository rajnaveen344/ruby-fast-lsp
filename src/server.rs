use crate::analyzer_prism::visitors::index_visitor::IndexVisitor;
use crate::handlers::{notification, request};
use crate::indexer::index::RubyIndex;
use crate::types::ruby_document::RubyDocument;
use anyhow::Result;
use log::{debug, info};
use lsp_types::{
    CompletionParams, CompletionResponse, DidChangeTextDocumentParams, DidCloseTextDocumentParams,
    DidOpenTextDocumentParams, GotoDefinitionParams, GotoDefinitionResponse, InitializeParams,
    InitializeResult, InitializedParams, InlayHintParams, Location, ReferenceParams,
    SemanticTokensParams, SemanticTokensResult, Url,
};
use ruby_prism::Visit;
use std::collections::HashMap;
use std::sync::{Arc, Mutex};
use std::time::Instant;
use tower_lsp::jsonrpc::Result as LspResult;
use tower_lsp::{Client, LanguageServer};

#[derive(Clone)]
pub struct RubyLanguageServer {
    pub client: Option<Client>,
    pub index: Arc<Mutex<RubyIndex>>,
    pub docs: Arc<Mutex<HashMap<Url, RubyDocument>>>,
}

impl RubyLanguageServer {
    pub fn new(client: Client) -> Result<Self> {
        let index = RubyIndex::new();
        Ok(Self {
            client: Some(client),
            index: Arc::new(Mutex::new(index)),
            docs: Arc::new(Mutex::new(HashMap::new())),
        })
    }

    pub fn index(&self) -> Arc<Mutex<RubyIndex>> {
        self.index.clone()
    }

    pub fn get_doc(&self, uri: &Url) -> Option<RubyDocument> {
        self.docs.lock().unwrap().get(uri).cloned()
    }

    pub fn process_file(&mut self, uri: Url, content: &str) -> Result<(), String> {
        self.index.lock().unwrap().remove_entries_for_uri(&uri);

        // Create or update document in the docs HashMap
        let document = RubyDocument::new(uri.clone(), content.to_string(), 0);
        self.docs.lock().unwrap().insert(uri.clone(), document);

        let parse_result = ruby_prism::parse(content.as_bytes());
        let node = parse_result.node();
        let mut visitor = IndexVisitor::new(self, uri.clone());

        visitor.visit(&node);

        // Persist mutations made by the visitor back to the server's document store
        // TODO: This is a temporary fix. We should be able to mutate the document in place
        //       using docs: Arc<Mutex<HashMap<Url, Arc<Mutex<RubyDocument>>>>>
        self.docs
            .lock()
            .unwrap()
            .insert(uri.clone(), visitor.document.clone());

        debug!("Processed file: {}", uri);
        Ok(())
    }
}

impl Default for RubyLanguageServer {
    fn default() -> Self {
        RubyLanguageServer {
            client: None,
            index: Arc::new(Mutex::new(RubyIndex::new())),
            docs: Arc::new(Mutex::new(HashMap::new())),
        }
    }
}

#[tower_lsp::async_trait]
impl LanguageServer for RubyLanguageServer {
    async fn initialize(&self, params: InitializeParams) -> LspResult<InitializeResult> {
        notification::handle_initialize(self, params).await
    }

    async fn initialized(&self, params: InitializedParams) {
        notification::handle_initialized(self, params).await
    }

    async fn shutdown(&self) -> LspResult<()> {
        notification::handle_shutdown(self).await
    }

    async fn did_open(&self, params: DidOpenTextDocumentParams) {
        info!("Document opened: {}", params.text_document.uri.path());
        let start_time = Instant::now();
        notification::handle_did_open(self, params).await;
        info!(
            "[PERF] Document open handler completed in {:?}",
            start_time.elapsed()
        );
    }

    async fn did_change(&self, params: DidChangeTextDocumentParams) {
        info!("Document changed: {}", params.text_document.uri.path());
        let start_time = Instant::now();
        notification::handle_did_change(self, params).await;
        info!(
            "[PERF] Document change handler completed in {:?}",
            start_time.elapsed()
        );
    }

    async fn did_close(&self, params: DidCloseTextDocumentParams) {
        info!("Document closed: {}", params.text_document.uri.path());
        let start_time = Instant::now();
        notification::handle_did_close(self, params).await;
        info!(
            "[PERF] Document close handler completed in {:?}",
            start_time.elapsed()
        );
    }

    async fn goto_definition(
        &self,
        params: GotoDefinitionParams,
    ) -> LspResult<Option<GotoDefinitionResponse>> {
        info!(
            "Goto definition request received for {:?}",
            params
                .text_document_position_params
                .text_document
                .uri
                .path()
        );
        let start_time = Instant::now();
        let result = request::handle_goto_definition(self, params).await;

        info!(
            "[PERF] Goto definition completed in {:?}",
            start_time.elapsed()
        );

        result
    }

    async fn references(&self, params: ReferenceParams) -> LspResult<Option<Vec<Location>>> {
        info!(
            "References request received for {:?}",
            params.text_document_position.text_document.uri.path()
        );
        let start_time = Instant::now();
        let result = request::handle_references(self, params).await;

        info!("[PERF] References completed in {:?}", start_time.elapsed());

        result
    }

    async fn semantic_tokens_full(
        &self,
        params: SemanticTokensParams,
    ) -> LspResult<Option<SemanticTokensResult>> {
        info!(
            "Semantic tokens request received for {:?}",
            params.text_document.uri.path()
        );
        let start_time = Instant::now();
        let result = request::handle_semantic_tokens_full(self, params).await;

        info!(
            "[PERF] Semantic tokens completed in {:?}",
            start_time.elapsed()
        );

        result
    }

    async fn inlay_hint(
        &self,
        params: InlayHintParams,
    ) -> LspResult<Option<Vec<lsp_types::InlayHint>>> {
        info!(
            "Inlay hint request received for {:?}",
            params.text_document.uri.path()
        );

        let start_time = Instant::now();
        let result = request::handle_inlay_hints(self, params).await;

        info!("[PERF] Inlay hint completed in {:?}", start_time.elapsed());

        result
    }

    async fn completion(&self, params: CompletionParams) -> LspResult<Option<CompletionResponse>> {
        info!(
            "Completion request received for {:?}",
            params.text_document_position.text_document.uri.path()
        );
        let start_time = Instant::now();
        let result = request::handle_completion(self, params).await;

        info!("[PERF] Completion completed in {:?}", start_time.elapsed());

        result
    }
}
