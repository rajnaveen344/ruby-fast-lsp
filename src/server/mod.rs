pub mod handlers;

use anyhow::Result;
use dashmap::DashMap;
use log::{info, warn};
use lsp_types::*;
use tower_lsp::jsonrpc::Result as LspResult;
use tower_lsp::{Client, LanguageServer};

use crate::parser::document::RubyDocument;
use self::handlers::RubyLspHandlers;
use std::sync::{Arc, Mutex};

pub struct RubyLanguageServer {
    pub client: Client,
    pub document_map: DashMap<Url, RubyDocument>,
    handlers: Option<Arc<Mutex<RubyLspHandlers>>>,
}

impl RubyLanguageServer {
    pub fn new(client: Client) -> Result<Self> {
        let handlers = Some(Arc::new(Mutex::new(RubyLspHandlers::new()?)));
        
        Ok(Self {
            client,
            document_map: DashMap::new(),
            handlers,
        })
    }
    
    pub fn new_fallback(client: Client) -> Self {
        warn!("Creating fallback Ruby LSP server with limited functionality");
        
        Self {
            client,
            document_map: DashMap::new(),
            handlers: None,
        }
    }
}

#[tower_lsp::async_trait]
impl LanguageServer for RubyLanguageServer {
    async fn initialize(&self, _: InitializeParams) -> LspResult<InitializeResult> {
        info!("Initializing Ruby LSP server");
        
        let capabilities = ServerCapabilities {
            text_document_sync: Some(TextDocumentSyncCapability::Kind(
                TextDocumentSyncKind::INCREMENTAL,
            )),
            hover_provider: Some(HoverProviderCapability::Simple(true)),
            completion_provider: Some(CompletionOptions {
                resolve_provider: Some(false),
                trigger_characters: Some(vec![".".to_string(), "::".to_string()]),
                work_done_progress_options: Default::default(),
                all_commit_characters: None,
                completion_item: None,
            }),
            definition_provider: Some(OneOf::Left(true)),
            // Add more capabilities as you implement them
            ..ServerCapabilities::default()
        };
        
        Ok(InitializeResult {
            capabilities,
            server_info: Some(ServerInfo {
                name: "ruby-fast-lsp".to_string(),
                version: Some(env!("CARGO_PKG_VERSION").to_string()),
            }),
        })
    }

    async fn initialized(&self, _: InitializedParams) {
        info!("Ruby LSP server initialized!");
        
        let message = if self.handlers.is_some() {
            "Ruby LSP server initialized with full functionality!"
        } else {
            "Ruby LSP server initialized with limited functionality. Some features may not work correctly."
        };
        
        self.client
            .log_message(MessageType::INFO, message)
            .await;
    }

    async fn shutdown(&self) -> LspResult<()> {
        info!("Shutting down Ruby LSP server");
        Ok(())
    }

    // Document synchronization methods
    async fn did_open(&self, params: DidOpenTextDocumentParams) {
        let uri = params.text_document.uri;
        let text = params.text_document.text;
        let version = params.text_document.version;
        
        info!("File opened: {}", uri);
        
        self.document_map.insert(
            uri.clone(),
            RubyDocument {
                content: text.clone(),
                version,
            },
        );
        
        self.client
            .log_message(MessageType::INFO, &format!("Opened document: {}", uri))
            .await;
    }

    async fn did_change(&self, params: DidChangeTextDocumentParams) {
        let uri = params.text_document.uri;
        let version = params.text_document.version;
        
        info!("File changed: {}", uri);
        
        if let Some(mut doc) = self.document_map.get_mut(&uri) {
            for change in params.content_changes {
                if let Some(_range) = change.range {
                    // Handle incremental updates
                    doc.content = change.text;
                } else {
                    // Full document update
                    doc.content = change.text;
                }
                doc.version = version;
            }
        }
    }

    async fn did_close(&self, params: DidCloseTextDocumentParams) {
        let uri = params.text_document.uri;
        
        info!("File closed: {}", uri);
        
        self.document_map.remove(&uri);
        
        self.client
            .log_message(MessageType::INFO, &format!("Closed document: {}", uri))
            .await;
    }

    // Language features
    async fn hover(&self, params: HoverParams) -> LspResult<Option<Hover>> {
        let uri = params.text_document_position_params.text_document.uri;
        let position = params.text_document_position_params.position;
        
        info!("Hover request at position {:?} in {}", position, uri);
        
        if let Some(handlers) = &self.handlers {
            if let Some(doc) = self.document_map.get(&uri) {
                if let Ok(handlers) = handlers.lock() {
                    return Ok(handlers.handle_hover(&doc, position));
                }
            }
        }
        
        // Fallback hover response
        Ok(Some(Hover {
            contents: HoverContents::Markup(MarkupContent {
                kind: MarkupKind::Markdown,
                value: "Ruby LSP running in fallback mode. Limited functionality available.".to_string(),
            }),
            range: None,
        }))
    }

    async fn completion(&self, params: CompletionParams) -> LspResult<Option<CompletionResponse>> {
        let uri = params.text_document_position.text_document.uri;
        let position = params.text_document_position.position;
        
        info!("Completion request at position {:?} in {}", position, uri);
        
        if let Some(handlers) = &self.handlers {
            if let Some(doc) = self.document_map.get(&uri) {
                if let Ok(handlers) = handlers.lock() {
                    return Ok(Some(handlers.handle_completion(&doc, position)));
                }
            }
        }
        
        // Fallback completion response with basic Ruby keywords
        let items = vec![
            CompletionItem {
                label: "def".to_string(),
                kind: Some(CompletionItemKind::KEYWORD),
                detail: Some("Define a method".to_string()),
                ..CompletionItem::default()
            },
            CompletionItem {
                label: "class".to_string(),
                kind: Some(CompletionItemKind::KEYWORD),
                detail: Some("Define a class".to_string()),
                ..CompletionItem::default()
            },
            CompletionItem {
                label: "module".to_string(),
                kind: Some(CompletionItemKind::KEYWORD),
                detail: Some("Define a module".to_string()),
                ..CompletionItem::default()
            },
        ];
        
        Ok(Some(CompletionResponse::Array(items)))
    }
    
    async fn goto_definition(
        &self,
        params: GotoDefinitionParams,
    ) -> LspResult<Option<GotoDefinitionResponse>> {
        let uri = params.text_document_position_params.text_document.uri;
        let position = params.text_document_position_params.position;
        
        info!("Definition request at position {:?} in {}", position, uri);
        
        if let Some(handlers) = &self.handlers {
            if let Some(doc) = self.document_map.get(&uri) {
                if let Ok(handlers) = handlers.lock() {
                    if let Some(range) = handlers.handle_definition(&doc, position) {
                        return Ok(Some(GotoDefinitionResponse::Scalar(Location {
                            uri: uri.clone(),
                            range,
                        })));
                    }
                }
            }
        }
        
        Ok(None)
    }
}
