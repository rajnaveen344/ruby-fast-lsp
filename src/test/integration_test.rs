//! Integration test harness – see `docs/integration_test_plan.md`
//!
//! This file purposely **replaces** the previous ad-hoc integration tests with a reusable
//! and extensible test harness.  The harness makes it straightforward to add new fixtures
//! and coverage as described in the integration test plan.

use std::path::{Path, PathBuf};
use std::sync::Once;

use lsp_types::{DidOpenTextDocumentParams, InitializeParams, TextDocumentItem, Url};
use tower_lsp::LanguageServer;

use crate::server::RubyLanguageServer;

/*----------------------------------------------------------------------
 Logger
----------------------------------------------------------------------*/

fn init_logger() {
    static INIT: Once = Once::new();
    INIT.call_once(|| {
        // Respect RUST_LOG env var but default to info for the test binary.
        env_logger::Builder::from_env(env_logger::Env::default().default_filter_or("debug"))
            .is_test(true)
            .init();
    });
}

/*----------------------------------------------------------------------
 Fixture helpers
----------------------------------------------------------------------*/

/// Absolute path to the root `fixtures/` directory.
fn fixture_root() -> PathBuf {
    // `CARGO_MANIFEST_DIR` resolves to the crate root at compile time.
    PathBuf::from(env!("CARGO_MANIFEST_DIR"))
        .join("src")
        .join("test")
        .join("fixtures")
}

/// Convert a [`Path`] to an [`Url`].  Panics if the conversion fails.
fn path_to_uri(path: &Path) -> Url {
    Url::from_file_path(path).expect("Failed to convert path to file:// URI")
}

/*----------------------------------------------------------------------
 TestHarness
----------------------------------------------------------------------*/

/// Encapsulates a `RubyLanguageServer` instance and helper utilities for
/// loading fixtures and performing common LSP requests.
pub struct TestHarness {
    server: RubyLanguageServer,
}

impl TestHarness {
    /// Create a fresh `RubyLanguageServer` instance and perform the
    /// `initialize` handshake.
    pub async fn new() -> Self {
        init_logger();

        let server = RubyLanguageServer::default();
        let _ = server.initialize(InitializeParams::default()).await;

        Self { server }
    }

    /// Opens **all** `*.rb` files located under `fixtures/<scenario>` so that
    /// the server indexes them as a workspace.
    pub async fn open_fixture_dir(&self, scenario: &str) {
        let root = fixture_root().join(scenario);
        assert!(root.exists(), "Unknown fixture scenario: {}", scenario);

        let mut stack = vec![root];
        while let Some(dir) = stack.pop() {
            for entry in std::fs::read_dir(&dir).expect("read_dir failed") {
                let entry = entry.expect("DirEntry failed");
                let path = entry.path();
                if path.is_dir() {
                    stack.push(path);
                } else if path.extension().and_then(|ext| ext.to_str()) == Some("rb") {
                    let uri = path_to_uri(&path);
                    let contents = std::fs::read_to_string(&path)
                        .unwrap_or_else(|_| panic!("Failed to read {:?}", path));

                    let params = DidOpenTextDocumentParams {
                        text_document: TextDocumentItem {
                            uri,
                            language_id: "ruby".into(),
                            version: 1,
                            text: contents,
                        },
                    };

                    // Fire-and-await so the document is fully indexed before continuing.
                    self.server.did_open(params).await;
                }
            }
        }
    }

    /// Borrow the underlying server – useful when calling handlers directly.
    pub fn server(&self) -> &RubyLanguageServer {
        &self.server
    }
}

/*----------------------------------------------------------------------
 Assertion helpers (macros)
----------------------------------------------------------------------*/

/// Assert that a *goto definition* request at (`file`, `line`, `char`) resolves
/// to (`exp_file`, `exp_line`).
#[macro_export]
macro_rules! assert_goto {
    ($harness:expr, $file:expr, $line:expr, $char:expr,
     $exp_file:expr, $exp_line:expr $(,)?) => {{
        use lsp_types::{
            GotoDefinitionParams, GotoDefinitionResponse, PartialResultParams, Position,
            TextDocumentIdentifier, TextDocumentPositionParams, WorkDoneProgressParams,
        };
        use $crate::handlers::request;

        let uri = path_to_uri(&fixture_root().join($file));
        let res = request::handle_goto_definition(
            $harness.server(),
            GotoDefinitionParams {
                text_document_position_params: TextDocumentPositionParams {
                    text_document: TextDocumentIdentifier { uri: uri.clone() },
                    position: Position {
                        line: $line,
                        character: $char,
                    },
                },
                work_done_progress_params: WorkDoneProgressParams::default(),
                partial_result_params: PartialResultParams::default(),
            },
        )
        .await
        .expect("goto definition request failed")
        .expect("no definition found");

        match res {
            GotoDefinitionResponse::Array(locations) => {
                assert_eq!(locations.len(), 1, "Expected exactly one location");
                let location = &locations[0];
                assert_eq!(location.uri, path_to_uri(&fixture_root().join($exp_file)));
                assert_eq!(location.range.start.line, $exp_line);
            }
            GotoDefinitionResponse::Scalar(location) => {
                // Fallback for unexpected scalar behaviour.
                assert_eq!(location.uri, path_to_uri(&fixture_root().join($exp_file)));
                assert_eq!(location.range.start.line, $exp_line);
            }
            other => panic!("Unexpected goto definition response: {:?}", other),
        }
    }};
}

/*----------------------------------------------------------------------
 Snapshot helpers
----------------------------------------------------------------------*/

/// Capture the definition locations at (`file`, `line`, `char`) and snapshot
/// the JSON array so it is easy to review when behaviour changes.
async fn snapshot_definitions(
    harness: &TestHarness,
    file: &str,
    line: u32,
    character: u32,
    snapshot_name: &str,
) {
    use crate::handlers::request;
    use lsp_types::{
        GotoDefinitionParams, PartialResultParams, Position, TextDocumentIdentifier,
        TextDocumentPositionParams, WorkDoneProgressParams,
    };

    let uri = path_to_uri(&fixture_root().join(file));
    let res = request::handle_goto_definition(
        harness.server(),
        GotoDefinitionParams {
            text_document_position_params: TextDocumentPositionParams {
                text_document: TextDocumentIdentifier { uri: uri.clone() },
                position: Position { line, character },
            },
            work_done_progress_params: WorkDoneProgressParams::default(),
            partial_result_params: PartialResultParams::default(),
        },
    )
    .await
    .expect("goto definition failed")
    .expect("no definition found");

    use serde_json::Value;
    let mut value: Value = match res {
        lsp_types::GotoDefinitionResponse::Array(loc) => serde_json::to_value(&loc).unwrap(),
        lsp_types::GotoDefinitionResponse::Scalar(l) => serde_json::to_value(&vec![l]).unwrap(),
        lsp_types::GotoDefinitionResponse::Link(ls) => serde_json::to_value(&ls).unwrap(),
    };

    // Replace absolute URIs with project-relative ones so snapshots are machine-independent
    fn normalize_uris(v: &mut Value, root_uri_prefix: &str) {
        match v {
            Value::Object(map) => {
                if let Some(Value::String(s)) = map.get_mut("uri") {
                    if s.starts_with(root_uri_prefix) {
                        let rel = &s[root_uri_prefix.len()..];
                        *s = format!("file://$PROJECT_ROOT/{}", rel);
                    }
                }
                for val in map.values_mut() {
                    normalize_uris(val, root_uri_prefix);
                }
            }
            Value::Array(arr) => {
                for val in arr {
                    normalize_uris(val, root_uri_prefix);
                }
            }
            _ => {}
        }
    }

    let project_root = std::env::var("CARGO_MANIFEST_DIR").unwrap();
    let root_uri_prefix = format!("file://{}/", project_root);
    normalize_uris(&mut value, &root_uri_prefix.replace("\\", "/"));

    insta::assert_json_snapshot!(snapshot_name, value);
}

/*----------------------------------------------------------------------
 Smoke test – validates that the harness itself works.
----------------------------------------------------------------------*/

#[cfg(test)]
mod tests {
    use super::*;

    /// Simply verifies that we can load a fixture directory without panicking.
    #[tokio::test]
    async fn harness_smoke() {
        let harness = TestHarness::new().await;
        // The `def_ref/` folder will be created in a follow-up commit.
        // For now fall back to an empty dir if it doesn't exist to keep CI green.
        if fixture_root().join("def_ref").exists() {
            harness.open_fixture_dir("def_ref").await;
        }
        // If we reached here, the harness is functional.
        assert!(true);
    }

    /// Validate definitions for module, class and constant in def_ref/single_file fixture.
    #[tokio::test]
    async fn def_ref_single_file_defs() {
        let harness = TestHarness::new().await;
        harness.open_fixture_dir("def_ref/single_file").await;

        // MyMod::Foo reference → class definition
        snapshot_definitions(
            &harness,
            "def_ref/single_file/single.rb",
            12,
            14,
            "foo_class_defs",
        )
        .await;

        // include MyMod → module definition
        snapshot_definitions(
            &harness,
            "def_ref/single_file/single.rb",
            10,
            8,
            "module_defs",
        )
        .await;

        // VALUE constant usage inside method → constant definition
        snapshot_definitions(
            &harness,
            "def_ref/single_file/single.rb",
            5,
            6,
            "value_constant_defs",
        )
        .await;

        // puts MyMod::VALUE constant usage at top level
        snapshot_definitions(
            &harness,
            "def_ref/single_file/single.rb",
            13,
            12,
            "value_constant_defs_top",
        )
        .await;
    }
}
