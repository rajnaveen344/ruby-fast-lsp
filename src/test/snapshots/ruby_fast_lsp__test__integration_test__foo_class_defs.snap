---
source: src/test/integration_test.rs
expression: value
---
[
  {
    "range": {
      "end": {
        "character": 5,
        "line": 7
      },
      "start": {
        "character": 2,
        "line": 3
      }
    },
    "uri": "file://$PROJECT_ROOT/src/test/fixtures/def_ref/single_file/single.rb"
  }
]
